llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1681,6 +1687,14 @@ bool MemCpyOptPass::performStackMoveOptzn(Instruction *Load, Instruction *Store,
   for (Instruction *I : NoAliasInstrs)
     I->setMetadata(LLVMContext::MD_noalias, nullptr);
 
+  // If we merge two allocas we need to uniform alias tags as well
+  if (!SrcAllocaInstUsers.empty()) {
+    MDNode *mergeTBAA =
+        (*SrcAllocaInstUsers.begin())->getMetadata(LLVMContext::MD_tbaa);
+    for (Instruction *It : DestAllocaInstUsers)
+      It->setMetadata(LLVMContext::MD_tbaa, mergeTBAA);
```

Comment:
This doesn't look safe? You are just assigning one tag to all the accesses now. There may be accesses with different tags, and replacing them with a different one would be a miscompile.

The easy fix here would be to just drop all the tbaa data like is done for noalias metadata. Otherwise you need to do by-offset analysis of accesses and merge the tbaa tags for specific offsets.

Reply:
In any case, we should have a test case where multiple tbaa tags are involved at different offsets.

Reply:
I dropped tbaa metadata and I added another test.